### ISSUE TITLE
Possible bug in Fast Gradient Method (minimal variant)

### ISSUE THREAD

**Issue by marcoancona** [IBM/adversarial-robustness-toolbox#204]
**Describe the bug**
There seems to be a bug in the implementation of Fast Gradient Method, in the minimal perturbation variant. In particular, the original input `x` is modified by the `generate` method. See the following example:

```python
import numpy as np
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential, Model
from tensorflow.keras.layers import Dense, Dropout, Flatten, Lambda, Activation, AveragePooling2D
from tensorflow.keras.layers import Conv2D, MaxPooling2D
from tensorflow.keras import backend as K
from evaluation import run_evaluation
from datetime import datetime
from utils.attacks_config import fast_attacks, test_attacks


print (f'Using Tensorflow {tf.__version__}')
print (f'Using Keras {keras.__version__}')


# input image dimensions
img_rows, img_cols = (28, 28)

# the data, split between train and test sets
(x_train, y_train), (x_test, y_test) = mnist.load_data()

x_train = x_train.reshape(x_train.shape[0], img_rows, img_cols, 1)
x_test = x_test.reshape(x_test.shape[0], img_rows, img_cols, 1)
input_shape = (img_rows, img_cols, 1)

x_train = x_train.astype("float32")
x_test = x_test.astype("float32")
x_train /= 255
x_test /= 255
x_train = (x_train - 0.5) * 2
x_test = (x_test - 0.5) * 2

# convert class vectors to binary class matrices
y_train = keras.utils.to_categorical(y_train, 10)
y_test = keras.utils.to_categorical(y_test, 10)


model = Sequential()
model.add(Flatten(input_shape=input_shape))
model.add(Dense(128))
model.add(Activation("relu"))
model.add(Dense(128))
model.add(Activation("relu"))
model.add(Dense(10))


def categorical_crossentropy(y_true, y_pred):
    return K.categorical_crossentropy(y_true, y_pred, from_logits=True)


model.compile(loss=categorical_crossentropy,
              optimizer=keras.optimizers.Adam(),
              metrics=['accuracy'])

model.fit(x_train, y_train,
          batch_size=128,
          epochs=1,
          verbose=1,
         )


x = np.copy(x_test[:100])
print ("This is zero: ", np.mean((x - x_test[:100])))  # < 0.0, of course

from art.classifiers import KerasClassifier
from art.attacks.fast_gradient import FastGradientMethod
from custom_attacks.ShapleyValuesMethod import ShapleyValuesMethod

classifier = KerasClassifier(
    model,
    use_logits=True,
    clip_values=(np.min(x_test), np.max(x_test))
)

attack = FastGradientMethod(classifier,
                            batch_size=1,
                            eps=5, # < using large values for the sake of the example
                            eps_step=1,
                            minimal=False)

x_adv = attack.generate(x)
print ("This should still be zero: ", np.mean((x - x_test[:100])))  # < 0.0, of course

attack = FastGradientMethod(classifier,
                            batch_size=1,
                            eps=5, # < using large values for the sake of the example
                            eps_step=1,
                            minimal=True)

x_adv = attack.generate(x)
print ("This should still be zero but it is not: ", np.mean((x - x_test[:100])))
```

Output
```
This is zero:  0.0
This should still be zero:  0.0
This should still be zero but it is not:  0.004872449
```

The problem is caused by the `_minimal_perturbation()` method. Notice that `x` is copied but then `_apply_perturbation()` is called on the original `x`, not on `adv_x`.
What is more, there is another connected issue: `_apply_perturbation()` is called with `current_eps` (and not `eps_step`), but since the original input is modified each time, the perturbation applied is much larger than it should.

```python
def _minimal_perturbation(self, x, y):
        """Iteratively compute the minimal perturbation necessary to make the class prediction change. Stop when the
        first adversarial example was found.
        :param x: An array with the original inputs
        :type x: `np.ndarray`
        :param y: Target values (class labels) one-hot-encoded of shape (nb_samples, nb_classes)
        :type y: `np.ndarray`
        :return: An array holding the adversarial examples
        :rtype: `np.ndarray`
        """
        adv_x = x.copy()

        # Compute perturbation with implicit batching
        for batch_id in range(int(np.ceil(adv_x.shape[0] / float(self.batch_size)))):
            batch_index_1, batch_index_2 = batch_id * self.batch_size, (batch_id + 1) * self.batch_size
            batch = adv_x[batch_index_1:batch_index_2]
            batch_labels = y[batch_index_1:batch_index_2]

            # Get perturbation
            perturbation = self._compute_perturbation(batch, batch_labels)

            # Get current predictions
            active_indices = np.arange(len(batch))
            current_eps = self.eps_step
            while active_indices.size > 0 and current_eps <= self.eps:
                # Adversarial crafting
                current_x = self._apply_perturbation(x[batch_index_1:batch_index_2], perturbation, current_eps)
                # Update
                batch[active_indices] = current_x[active_indices]
                adv_preds = self.classifier.predict(batch)
                # If targeted active check to see whether we have hit the target, otherwise head to anything but
                if self.targeted:
                    active_indices = np.where(np.argmax(batch_labels, axis=1) != np.argmax(adv_preds, axis=1))[0]
                else:
                    active_indices = np.where(np.argmax(batch_labels, axis=1) == np.argmax(adv_preds, axis=1))[0]

                current_eps += self.eps_step

            adv_x[batch_index_1:batch_index_2] = batch

        return adv_x
```

**Comment by beat-buesser** [comment]
Hi @marcoancona Thank you for raising this issue. Yes, the intention is that the attacks should not modify the original input.

@minhitbk What do you think? Should line 113 below https://github.com/IBM/adversarial-robustness-toolbox/blob/548febd02770bf06d9e0bb34974b3d98ec889865/art/attacks/fast_gradient.py#L113

be changed to 
```
current_x = self._apply_perturbation(adv_x[batch_index_1:batch_index_2], perturbation, current_eps)
```

I think this does not affect the quality of generated adversarial samples.

**Comment by beat-buesser** [comment]
@marcoancona Would you be interested in implementing the possible solution and make a pull request for it?

**Comment by imolloy** [comment]
I’m not sure if you could replace `x` with `adv_x` depending on how `active_indices` is updated. What about changing `_apply_perturbation`, specifically this line: https://github.com/IBM/adversarial-robustness-toolbox/blob/548febd02770bf06d9e0bb34974b3d98ec889865/art/attacks/fast_gradient.py#L252

**Comment by minhitbk** [comment]
I agree with @imolloy, changing x to adv_x in line 113 won't work because adv_x will again be changed after one iteration in the while loop. Instead, the root cause is in line 251 as @imolloy specified. So we should change
    batch += eps_step * perturbation
to 
    batch = batch + eps_step * perturbation
This solution should work.

**Comment by beat-buesser** [comment]
Thank you @minhitbk and @imolloy It works for me using the script above.

@marcoancona What do you think?

**Comment by marcoancona** [comment]
Yes, the change suggested by @minhitbk should work. And since `batch` in this case is always the original input, it is correct to call `_apply_perturbation` with `current_eps`
